code/modules/overmap/ships/computers/comms.dm

```diff
@@ -0,0 +1,306 @@
+		entities += list(list(
+			"name" = "None.",
+			"coords" = "-,-",
+			"desc" = "-",
+			"status" = "No local entities responding to comms ping."
+		))
+	if(!length(broadcasters))
+		broadcasters += list(list(
+			"name" = "None.",
+			"coords" = "-,-,-",
+			"desc" = "No broadcasters available."
+		))
+	if(!length(receivers))
+		receivers += list(list(
+			"name" = "None.",
+			"coords" = "-,-,-",
+			"desc" = "No receivers available."
+		))
+
+	data["allow_ident_change"] = linked.can_switch_ident
+	data["ident_enabled"] =      linked.ident_transmitter
+
+	data["entities"] =     entities
+	data["broadcasters"] = broadcasters
+	data["receivers"] =    receivers
+
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "shipcomms.tmpl", "[linked.name] Communication Screen", 380, 530)
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/machinery/computer/ship/comms/OnTopic(var/mob/user, var/list/href_list)
+	if(..())
+		return TOPIC_HANDLED
+	if (!linked)
+		return TOPIC_NOACTION
+	if(href_list["toggle_comms_visibility"])
+		if(linked.can_switch_ident)
+			linked.ident_transmitter = !linked.ident_transmitter
+			. = TOPIC_REFRESH
+
+/obj/machinery/computer/ship/comms/proc/is_destination_in_range(var/obj/effect/overmap/destination)
+	. = FALSE
+	for(var/obj/machinery/shipcomms/comms in destination.comms_masers)
+		var/functional = (comms.stat & (BROKEN|NOPOWER))
+		if(functional)
+			if(destination in comms.get_nearby_entitities())
+				. = TRUE
+
+/obj/machinery/computer/ship/comms/proc/has_antennae_functional(var/obj/effect/overmap/overmap_obj)
+	. = FALSE
+	for(var/obj/machinery/shipcomms/comms in overmap_obj.comms_antennae)
+		if((comms.stat & (BROKEN|NOPOWER)))
+			. = TRUE
+			break
+
+/obj/machinery/computer/ship/comms/proc/send_message(var/user, var/obj/effect/overmap/visitable/destination, var/message)
+	if(is_destination_in_range(destination))
+		for(var/obj/machinery/computer/ship/comms/C in destination.associated_machinery)
+			if(has_antennae_functional(destination))
+				C.recieve_message(linked, message)
+				ping("Message sent.")
+				break
+			else
+				to_chat(user, SPAN_WARNING("No confirmation ping recieved - message sent but not recieved."))
+				break
+	else
+		to_chat(user, SPAN_WARNING("Destination node out of range."))
+		return
+
+/obj/machinery/computer/ship/comms/proc/recieve_message(var/obj/effect/overmap/source, var/message)
+	var/datum/communication/comm_log
+	for(var/datum/communication/C in communications) //Check if we have an existing communications log with the source.
+		if(C.source == source)
+			comm_log = C
+	comm_log.add_message_to_log(source, message)
+	ping("Message recieved from [source.ident_transmitter ? source.name : "Unknown"].")
+
+/obj/machinery/computer/ship/comms/proc/open_comms(var/destination)
+	if(has_antennae_functional(destination))
+		for(var/obj/machinery/computer/ship/comms/C in destination.associated_machinery)
```

Comment:
`for(var/obj/machinery/computer/ship/comms/C as anything in in destination.get_linked_machines_of_type(/obj/machinery/computer/ship/comms))`